#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
gRPC server for auth-service.
OAuth 2.0 è®¤è¯æœåŠ¡
"""

from __future__ import annotations

import json
import os
import sys
import time
import secrets
from concurrent import futures
from datetime import datetime, timedelta
from typing import Optional, Dict, Any

import grpc

BASE_DIR = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
PROJECT_ROOT = BASE_DIR
sys.path.insert(0, PROJECT_ROOT)

# å¯¼å…¥ç”Ÿæˆçš„ gRPC ä»£ç 
sys.path.insert(0, os.path.join(PROJECT_ROOT, "proto", "generated"))
import auth_pb2
import auth_pb2_grpc

# å¯¼å…¥é…ç½®
from .config import (
    REDIS_HOST, REDIS_PORT, REDIS_DB, REDIS_PASSWORD,
    OAUTH_CLIENT_ID, OAUTH_CLIENT_SECRET,
    OAUTH_ACCESS_TOKEN_EXPIRE_MINUTES,
    OAUTH_REFRESH_TOKEN_EXPIRE_DAYS,
    REDIS_KEY_PREFIX_ACCESS_TOKEN,
    REDIS_KEY_PREFIX_REFRESH_TOKEN,
    REDIS_KEY_PREFIX_AUTH_CODE,
    REDIS_KEY_PREFIX_TOKEN_INFO,
)

# å°è¯•å¯¼å…¥ Redis
try:
    import redis
    REDIS_AVAILABLE = True
except ImportError:
    REDIS_AVAILABLE = False
    print("âš ï¸  Redis æœªå®‰è£…ï¼Œå°†ä½¿ç”¨å†…å­˜å­˜å‚¨ï¼ˆä»…ç”¨äºå¼€å‘ï¼‰", flush=True)


class AuthServicer(auth_pb2_grpc.AuthServiceServicer):
    """å®ç° AuthService çš„ gRPC æœåŠ¡"""

    def __init__(self):
        """åˆå§‹åŒ–è®¤è¯æœåŠ¡"""
        self.redis_client = None
        if REDIS_AVAILABLE:
            try:
                self.redis_client = redis.Redis(
                    host=REDIS_HOST,
                    port=REDIS_PORT,
                    db=REDIS_DB,
                    password=REDIS_PASSWORD,
                    decode_responses=True,
                    socket_connect_timeout=2
                )
                # æµ‹è¯•è¿æ¥
                self.redis_client.ping()
                print(f"âœ“ Redis è¿æ¥æˆåŠŸ: {REDIS_HOST}:{REDIS_PORT}", flush=True)
            except Exception as e:
                print(f"âš ï¸  Redis è¿æ¥å¤±è´¥: {e}ï¼Œå°†ä½¿ç”¨å†…å­˜å­˜å‚¨", flush=True)
                self.redis_client = None
        
        # å†…å­˜å­˜å‚¨ï¼ˆRedis ä¸å¯ç”¨æ—¶çš„é™çº§æ–¹æ¡ˆï¼‰
        self._memory_storage = {}
        self._memory_storage_ttl = {}

    def _get_redis_key(self, key_type: str, token: str) -> str:
        """è·å– Redis é”®"""
        prefixes = {
            "access_token": REDIS_KEY_PREFIX_ACCESS_TOKEN,
            "refresh_token": REDIS_KEY_PREFIX_REFRESH_TOKEN,
            "auth_code": REDIS_KEY_PREFIX_AUTH_CODE,
            "token_info": REDIS_KEY_PREFIX_TOKEN_INFO,
        }
        return f"{prefixes.get(key_type, 'auth:')}{token}"

    def _get_token_info(self, token: str) -> Optional[Dict[str, Any]]:
        """ä»å­˜å‚¨ä¸­è·å– Token ä¿¡æ¯"""
        token_prefix = token[:20] + "..." if len(token) > 20 else token
        
        if self.redis_client:
            try:
                key = self._get_redis_key("token_info", token)
                data = self.redis_client.get(key)
                if data:
                    info = json.loads(data)
                    print(f"    âœ“ Token ä¿¡æ¯ä» Redis è·å–: {token_prefix}", flush=True)
                    return info
                else:
                    print(f"    âœ— Token ä¿¡æ¯åœ¨ Redis ä¸­ä¸å­˜åœ¨: {token_prefix} (key: {key})", flush=True)
            except Exception as e:
                print(f"    âš ï¸  Redis è¯»å–å¤±è´¥: {e}", flush=True)
        
        # é™çº§åˆ°å†…å­˜å­˜å‚¨
        if token in self._memory_storage:
            # æ£€æŸ¥æ˜¯å¦è¿‡æœŸ
            if token in self._memory_storage_ttl:
                if time.time() > self._memory_storage_ttl[token]:
                    print(f"    âœ— Token ä¿¡æ¯åœ¨å†…å­˜ä¸­å·²è¿‡æœŸ: {token_prefix}", flush=True)
                    del self._memory_storage[token]
                    del self._memory_storage_ttl[token]
                    return None
            print(f"    âœ“ Token ä¿¡æ¯ä»å†…å­˜è·å–: {token_prefix}", flush=True)
            return self._memory_storage[token]
        
        print(f"    âœ— Token ä¿¡æ¯ä¸å­˜åœ¨: {token_prefix}", flush=True)
        return None

    def _save_token_info(self, token: str, info: Dict[str, Any], ttl_seconds: int):
        """ä¿å­˜ Token ä¿¡æ¯åˆ°å­˜å‚¨"""
        if self.redis_client:
            try:
                key = self._get_redis_key("token_info", token)
                self.redis_client.setex(key, ttl_seconds, json.dumps(info, ensure_ascii=False))
            except Exception as e:
                print(f"âš ï¸  Redis å†™å…¥å¤±è´¥: {e}", flush=True)
        
        # é™çº§åˆ°å†…å­˜å­˜å‚¨
        self._memory_storage[token] = info
        self._memory_storage_ttl[token] = time.time() + ttl_seconds

    def _check_token_exists(self, token: str, token_type: str = "access_token") -> bool:
        """æ£€æŸ¥ Token æ˜¯å¦å­˜åœ¨ä¸”æœ‰æ•ˆ"""
        token_prefix = token[:20] + "..." if len(token) > 20 else token
        
        if self.redis_client:
            try:
                key = self._get_redis_key(token_type, token)
                exists = self.redis_client.exists(key) > 0
                if exists:
                    print(f"    âœ“ Token åœ¨ Redis ä¸­æ‰¾åˆ°: {token_prefix} (key: {key})", flush=True)
                else:
                    print(f"    âœ— Token åœ¨ Redis ä¸­ä¸å­˜åœ¨: {token_prefix} (key: {key})", flush=True)
                return exists
            except Exception as e:
                print(f"    âš ï¸  Redis æŸ¥è¯¢å¤±è´¥: {e}", flush=True)
        
        # é™çº§åˆ°å†…å­˜å­˜å‚¨
        if token in self._memory_storage:
            if token in self._memory_storage_ttl:
                if time.time() > self._memory_storage_ttl[token]:
                    print(f"    âœ— Token åœ¨å†…å­˜ä¸­å·²è¿‡æœŸ: {token_prefix}", flush=True)
                    del self._memory_storage[token]
                    del self._memory_storage_ttl[token]
                    return False
            print(f"    âœ“ Token åœ¨å†…å­˜ä¸­æ‰¾åˆ°: {token_prefix}", flush=True)
            return True
        
        print(f"    âœ— Token åœ¨å†…å­˜ä¸­ä¸å­˜åœ¨: {token_prefix}", flush=True)
        return False

    def _verify_jwt_token(self, token: str) -> Optional[Dict[str, Any]]:
        """éªŒè¯ JWT Tokenï¼ˆå‘åå…¼å®¹ï¼‰"""
        try:
            import jwt
            import os
            from datetime import datetime, timezone
            
            # è·å– JWT Secret
            secret = os.getenv("JWT_SECRET") or "dev-secret-change-me"
            
            # è§£ç  JWT Token
            payload = jwt.decode(token, secret, algorithms=["HS256"])
            
            # æ£€æŸ¥æ˜¯å¦è¿‡æœŸ
            exp = payload.get("exp")
            if exp:
                current_timestamp = datetime.now(timezone.utc).timestamp()
                if current_timestamp > exp:
                    print(f"    âœ— JWT Token å·²è¿‡æœŸ (exp: {exp}, current: {current_timestamp})", flush=True)
                    return None  # Token å·²è¿‡æœŸ
            
            # è¿”å›ç”¨æˆ·ä¿¡æ¯
            user_id = payload.get("sub", "")
            print(f"    âœ“ JWT Token éªŒè¯æˆåŠŸ (user_id: {user_id}, exp: {exp})", flush=True)
            return {
                "user_id": user_id,
                "client_id": "jwt_client",
                "expires_at": exp if exp else 0,
                "issued_at": payload.get("iat", 0),
                "scope": []
            }
        except jwt.ExpiredSignatureError as e:
            print(f"    âœ— JWT Token å·²è¿‡æœŸ: {e}", flush=True)
            return None
        except jwt.InvalidTokenError as e:
            print(f"    âœ— JWT Token æ— æ•ˆ: {e}", flush=True)
            return None
        except ImportError as e:
            print(f"    âš ï¸  PyJWT æœªå®‰è£…ï¼Œæ— æ³•éªŒè¯ JWT Token: {e}", flush=True)
            return None
        except Exception as e:
            print(f"    âœ— JWT Token éªŒè¯å¼‚å¸¸: {e}", flush=True)
            return None
    
    def VerifyToken(self, request: auth_pb2.VerifyTokenRequest, context: grpc.ServicerContext) -> auth_pb2.VerifyTokenResponse:
        """éªŒè¯ Token æ˜¯å¦æœ‰æ•ˆï¼ˆæ”¯æŒ JWT å’Œ OAuth Tokenï¼‰"""
        import datetime
        request_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        
        try:
            token = request.token
            if not token:
                print(f"[{request_time}] âŒ auth-service: Token ä¸ºç©º", flush=True)
                return auth_pb2.VerifyTokenResponse(
                    valid=False,
                    error="Token ä¸ºç©º"
                )
            
            # è®°å½• Token å‰ç¼€ï¼ˆç”¨äºè°ƒè¯•ï¼Œä¸å®Œæ•´æ˜¾ç¤ºï¼‰
            token_prefix = token[:20] + "..." if len(token) > 20 else token
            print(f"[{request_time}] ğŸ“¥ auth-service: éªŒè¯ Token ({token_prefix})", flush=True)
            
            # é¦–å…ˆå°è¯•ä½œä¸º OAuth Token éªŒè¯
            print(f"    å°è¯•ä½œä¸º OAuth Token éªŒè¯...", flush=True)
            redis_status = "å¯ç”¨" if self.redis_client else "ä¸å¯ç”¨ï¼ˆä½¿ç”¨å†…å­˜å­˜å‚¨ï¼‰"
            print(f"    Redis çŠ¶æ€: {redis_status}", flush=True)
            
            # æ£€æŸ¥ Access Token æ˜¯å¦å­˜åœ¨
            if self._check_token_exists(token, "access_token"):
                # è·å– Token ä¿¡æ¯
                print(f"    è·å– OAuth Token ä¿¡æ¯...", flush=True)
                token_info = self._get_token_info(token)
                if token_info:
                    # æ£€æŸ¥æ˜¯å¦è¿‡æœŸ
                    expires_at = token_info.get("expires_at", 0)
                    current_time = time.time()
                    if expires_at > 0:
                        if current_time > expires_at:
                            remaining = expires_at - current_time
                            print(f"[{request_time}] âŒ auth-service: OAuth Token å·²è¿‡æœŸ ({token_prefix}, è¿‡æœŸæ—¶é—´: {expires_at}, å½“å‰æ—¶é—´: {current_time}, å‰©ä½™: {remaining}ç§’)", flush=True)
                            return auth_pb2.VerifyTokenResponse(
                                valid=False,
                                error="Token å·²è¿‡æœŸ"
                            )
                        else:
                            remaining = expires_at - current_time
                            print(f"    OAuth Token æœ‰æ•ˆæœŸå‰©ä½™: {remaining:.0f} ç§’", flush=True)
                    
                    print(f"[{request_time}] âœ… auth-service: OAuth Token éªŒè¯æˆåŠŸ ({token_prefix}, user_id: {token_info.get('user_id', 'N/A')})", flush=True)
                    return auth_pb2.VerifyTokenResponse(
                        valid=True,
                        user_id=token_info.get("user_id", ""),
                        client_id=token_info.get("client_id", ""),
                        expires_at=int(expires_at)
                    )
            
            # å¦‚æœ OAuth Token éªŒè¯å¤±è´¥ï¼Œå°è¯•ä½œä¸º JWT Token éªŒè¯ï¼ˆå‘åå…¼å®¹ï¼‰
            print(f"    OAuth Token éªŒè¯å¤±è´¥ï¼Œå°è¯•ä½œä¸º JWT Token éªŒè¯...", flush=True)
            jwt_info = self._verify_jwt_token(token)
            if jwt_info:
                expires_at = jwt_info.get("expires_at", 0)
                print(f"[{request_time}] âœ… auth-service: JWT Token éªŒè¯æˆåŠŸ ({token_prefix}, user_id: {jwt_info.get('user_id', 'N/A')})", flush=True)
                return auth_pb2.VerifyTokenResponse(
                    valid=True,
                    user_id=jwt_info.get("user_id", ""),
                    client_id=jwt_info.get("client_id", ""),
                    expires_at=int(expires_at)
                )
            
            # ä¸¤ç§éªŒè¯éƒ½å¤±è´¥
            print(f"[{request_time}] âŒ auth-service: Token ä¸å­˜åœ¨æˆ–å·²è¿‡æœŸ ({token_prefix})", flush=True)
            return auth_pb2.VerifyTokenResponse(
                valid=False,
                error="Token ä¸å­˜åœ¨æˆ–å·²è¿‡æœŸ"
            )
            
        except Exception as e:
            import traceback
            error_msg = f"éªŒè¯å¤±è´¥: {str(e)}\n{traceback.format_exc()}"
            print(f"[{request_time}] âŒ auth-service: é”™è¯¯ - {error_msg}", flush=True)
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"éªŒè¯å¤±è´¥: {str(e)}")
            return auth_pb2.VerifyTokenResponse(
                valid=False,
                error=f"éªŒè¯å¤±è´¥: {str(e)}"
            )

    def RefreshToken(self, request: auth_pb2.RefreshTokenRequest, context: grpc.ServicerContext) -> auth_pb2.RefreshTokenResponse:
        """åˆ·æ–° Token"""
        import datetime
        request_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        print(f"[{request_time}] ğŸ“¥ auth-service: åˆ·æ–° Token", flush=True)
        
        try:
            refresh_token = request.refresh_token
            client_id = request.client_id
            client_secret = request.client_secret
            
            # éªŒè¯å®¢æˆ·ç«¯å‡­è¯
            if client_id != OAUTH_CLIENT_ID or client_secret != OAUTH_CLIENT_SECRET:
                return auth_pb2.RefreshTokenResponse(
                    success=False,
                    error="å®¢æˆ·ç«¯å‡­è¯æ— æ•ˆ"
                )
            
            # æ£€æŸ¥ Refresh Token æ˜¯å¦å­˜åœ¨
            if not self._check_token_exists(refresh_token, "refresh_token"):
                return auth_pb2.RefreshTokenResponse(
                    success=False,
                    error="Refresh Token ä¸å­˜åœ¨æˆ–å·²è¿‡æœŸ"
                )
            
            # è·å– Refresh Token ä¿¡æ¯
            refresh_info = self._get_token_info(refresh_token)
            if not refresh_info:
                return auth_pb2.RefreshTokenResponse(
                    success=False,
                    error="Refresh Token ä¿¡æ¯ä¸å­˜åœ¨"
                )
            
            # æ£€æŸ¥æ˜¯å¦è¿‡æœŸ
            expires_at = refresh_info.get("expires_at", 0)
            if expires_at > 0 and time.time() > expires_at:
                return auth_pb2.RefreshTokenResponse(
                    success=False,
                    error="Refresh Token å·²è¿‡æœŸ"
                )
            
            # ç”Ÿæˆæ–°çš„ Access Token
            user_id = refresh_info.get("user_id", "")
            new_access_token = secrets.token_urlsafe(32)
            access_token_expires_at = time.time() + (OAUTH_ACCESS_TOKEN_EXPIRE_MINUTES * 60)
            
            # ä¿å­˜æ–°çš„ Access Token
            access_token_info = {
                "user_id": user_id,
                "client_id": client_id,
                "expires_at": access_token_expires_at,
                "issued_at": time.time()
            }
            self._save_token_info(new_access_token, access_token_info, OAUTH_ACCESS_TOKEN_EXPIRE_MINUTES * 60)
            
            # ä¿å­˜ Access Token åˆ° Redisï¼ˆç”¨äºå¿«é€ŸéªŒè¯ï¼‰
            if self.redis_client:
                try:
                    key = self._get_redis_key("access_token", new_access_token)
                    self.redis_client.setex(key, OAUTH_ACCESS_TOKEN_EXPIRE_MINUTES * 60, "1")
                except Exception:
                    pass
            
            print(f"[{request_time}] âœ… auth-service: Token åˆ·æ–°æˆåŠŸ", flush=True)
            return auth_pb2.RefreshTokenResponse(
                success=True,
                access_token=new_access_token,
                expires_in=OAUTH_ACCESS_TOKEN_EXPIRE_MINUTES * 60,
                token_type="bearer"
            )
            
        except Exception as e:
            import traceback
            error_msg = f"åˆ·æ–°å¤±è´¥: {str(e)}\n{traceback.format_exc()}"
            print(f"[{request_time}] âŒ auth-service: é”™è¯¯ - {error_msg}", flush=True)
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"åˆ·æ–°å¤±è´¥: {str(e)}")
            return auth_pb2.RefreshTokenResponse(
                success=False,
                error=f"åˆ·æ–°å¤±è´¥: {str(e)}"
            )

    def GetTokenInfo(self, request: auth_pb2.GetTokenInfoRequest, context: grpc.ServicerContext) -> auth_pb2.GetTokenInfoResponse:
        """è·å– Token ä¿¡æ¯"""
        import datetime
        request_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        
        try:
            token = request.token
            if not token:
                print(f"[{request_time}] âŒ auth-service: Token ä¸ºç©º", flush=True)
                return auth_pb2.GetTokenInfoResponse(
                    valid=False,
                    error="Token ä¸ºç©º"
                )
            
            token_prefix = token[:20] + "..." if len(token) > 20 else token
            print(f"[{request_time}] ğŸ“¥ auth-service: è·å– Token ä¿¡æ¯ ({token_prefix})", flush=True)
            
            # æ£€æŸ¥ Redis è¿æ¥çŠ¶æ€
            redis_status = "å¯ç”¨" if self.redis_client else "ä¸å¯ç”¨ï¼ˆä½¿ç”¨å†…å­˜å­˜å‚¨ï¼‰"
            print(f"    Redis çŠ¶æ€: {redis_status}", flush=True)
            
            # æ£€æŸ¥ Token æ˜¯å¦å­˜åœ¨
            print(f"    æ£€æŸ¥ Token æ˜¯å¦å­˜åœ¨...", flush=True)
            if not self._check_token_exists(token, "access_token"):
                print(f"[{request_time}] âŒ auth-service: Token ä¸å­˜åœ¨æˆ–å·²è¿‡æœŸ ({token_prefix})", flush=True)
                return auth_pb2.GetTokenInfoResponse(
                    valid=False,
                    error="Token ä¸å­˜åœ¨æˆ–å·²è¿‡æœŸ"
                )
            
            # è·å– Token ä¿¡æ¯
            print(f"    è·å– Token ä¿¡æ¯...", flush=True)
            token_info = self._get_token_info(token)
            if not token_info:
                print(f"[{request_time}] âŒ auth-service: Token ä¿¡æ¯ä¸å­˜åœ¨ ({token_prefix})", flush=True)
                return auth_pb2.GetTokenInfoResponse(
                    valid=False,
                    error="Token ä¿¡æ¯ä¸å­˜åœ¨"
                )
            
            # æ£€æŸ¥æ˜¯å¦è¿‡æœŸ
            expires_at = token_info.get("expires_at", 0)
            current_time = time.time()
            if expires_at > 0:
                if current_time > expires_at:
                    remaining = expires_at - current_time
                    print(f"[{request_time}] âŒ auth-service: Token å·²è¿‡æœŸ ({token_prefix}, è¿‡æœŸæ—¶é—´: {expires_at}, å½“å‰æ—¶é—´: {current_time}, å‰©ä½™: {remaining}ç§’)", flush=True)
                    return auth_pb2.GetTokenInfoResponse(
                        valid=False,
                        error="Token å·²è¿‡æœŸ"
                    )
                else:
                    remaining = expires_at - current_time
                    print(f"    Token æœ‰æ•ˆæœŸå‰©ä½™: {remaining:.0f} ç§’", flush=True)
            
            print(f"[{request_time}] âœ… auth-service: Token ä¿¡æ¯è·å–æˆåŠŸ ({token_prefix}, user_id: {token_info.get('user_id', 'N/A')})", flush=True)
            response = auth_pb2.GetTokenInfoResponse(
                valid=True,
                user_id=token_info.get("user_id", ""),
                client_id=token_info.get("client_id", ""),
                issued_at=int(token_info.get("issued_at", 0)),
                expires_at=int(expires_at)
            )
            # æ·»åŠ  scopeï¼ˆå¦‚æœæœ‰ï¼‰
            if "scope" in token_info:
                response.scope.extend(token_info["scope"])
            
            return response
            
        except Exception as e:
            import traceback
            error_msg = f"è·å–å¤±è´¥: {str(e)}\n{traceback.format_exc()}"
            print(f"[{request_time}] âŒ auth-service: é”™è¯¯ - {error_msg}", flush=True)
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"è·å–å¤±è´¥: {str(e)}")
            return auth_pb2.GetTokenInfoResponse(
                valid=False,
                error=f"è·å–å¤±è´¥: {str(e)}"
            )

    def CreateToken(self, request: auth_pb2.CreateTokenRequest, context: grpc.ServicerContext) -> auth_pb2.CreateTokenResponse:
        """åˆ›å»º Token"""
        import datetime
        request_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        print(f"[{request_time}] ğŸ“¥ auth-service: åˆ›å»º Token", flush=True)
        
        try:
            user_id = request.user_id
            client_id = request.client_id
            scope = list(request.scope) if request.scope else []
            
            # ç¡®å®šè¿‡æœŸæ—¶é—´ï¼ˆä½¿ç”¨è¯·æ±‚å€¼æˆ–é»˜è®¤å€¼ï¼‰
            access_token_expires_in = request.access_token_expires_in
            if access_token_expires_in <= 0:
                access_token_expires_in = OAUTH_ACCESS_TOKEN_EXPIRE_MINUTES * 60
            
            refresh_token_expires_in = request.refresh_token_expires_in
            if refresh_token_expires_in <= 0:
                refresh_token_expires_in = OAUTH_REFRESH_TOKEN_EXPIRE_DAYS * 24 * 60 * 60
            
            # ç”Ÿæˆ Access Token å’Œ Refresh Token
            access_token = secrets.token_urlsafe(32)
            refresh_token = secrets.token_urlsafe(32)
            
            # è®¡ç®—è¿‡æœŸæ—¶é—´æˆ³
            current_time = time.time()
            access_token_expires_at = current_time + access_token_expires_in
            refresh_token_expires_at = current_time + refresh_token_expires_in
            
            # ä¿å­˜ Access Token ä¿¡æ¯
            access_token_info = {
                "user_id": user_id,
                "client_id": client_id,
                "scope": scope,
                "expires_at": access_token_expires_at,
                "issued_at": current_time
            }
            self._save_token_info(access_token, access_token_info, access_token_expires_in)
            
            # ä¿å­˜ Access Token åˆ° Redisï¼ˆç”¨äºå¿«é€ŸéªŒè¯ï¼‰
            if self.redis_client:
                try:
                    key = self._get_redis_key("access_token", access_token)
                    self.redis_client.setex(key, access_token_expires_in, "1")
                except Exception as e:
                    print(f"âš ï¸  Redis ä¿å­˜ Access Token å¤±è´¥: {e}", flush=True)
            
            # ä¿å­˜ Refresh Token ä¿¡æ¯
            refresh_token_info = {
                "user_id": user_id,
                "client_id": client_id,
                "scope": scope,
                "expires_at": refresh_token_expires_at,
                "issued_at": current_time
            }
            self._save_token_info(refresh_token, refresh_token_info, refresh_token_expires_in)
            
            # ä¿å­˜ Refresh Token åˆ° Redisï¼ˆç”¨äºå¿«é€ŸéªŒè¯ï¼‰
            if self.redis_client:
                try:
                    key = self._get_redis_key("refresh_token", refresh_token)
                    self.redis_client.setex(key, refresh_token_expires_in, "1")
                except Exception as e:
                    print(f"âš ï¸  Redis ä¿å­˜ Refresh Token å¤±è´¥: {e}", flush=True)
            
            print(f"[{request_time}] âœ… auth-service: Token åˆ›å»ºæˆåŠŸ", flush=True)
            return auth_pb2.CreateTokenResponse(
                success=True,
                access_token=access_token,
                refresh_token=refresh_token,
                expires_in=access_token_expires_in,
                token_type="bearer"
            )
            
        except Exception as e:
            import traceback
            error_msg = f"åˆ›å»ºå¤±è´¥: {str(e)}\n{traceback.format_exc()}"
            print(f"[{request_time}] âŒ auth-service: é”™è¯¯ - {error_msg}", flush=True)
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"åˆ›å»ºå¤±è´¥: {str(e)}")
            return auth_pb2.CreateTokenResponse(
                success=False,
                error=f"åˆ›å»ºå¤±è´¥: {str(e)}"
            )

    def HealthCheck(self, request: auth_pb2.HealthCheckRequest, context: grpc.ServicerContext) -> auth_pb2.HealthCheckResponse:
        """å¥åº·æ£€æŸ¥"""
        return auth_pb2.HealthCheckResponse(status="ok")


def serve(port: int = 9011):
    """å¯åŠ¨ gRPC æœåŠ¡å™¨ï¼ˆæ”¯æŒçƒ­æ›´æ–°ï¼‰"""
    try:
        # å°è¯•ä½¿ç”¨çƒ­æ›´æ–°æ¨¡å¼
        from server.hot_reload.microservice_reloader import (
            create_hot_reload_server,
            register_microservice_reloader
        )
        
        # æœåŠ¡å™¨é€‰é¡¹
        server_options = [
            ('grpc.keepalive_time_ms', 300000),  # 5åˆ†é’Ÿ
            ('grpc.keepalive_timeout_ms', 20000),  # 20ç§’
            ('grpc.keepalive_permit_without_calls', False),
            ('grpc.http2.max_pings_without_data', 2),
            ('grpc.http2.min_time_between_pings_ms', 60000),  # 60ç§’
            ('grpc.http2.min_ping_interval_without_data_ms', 300000),  # 5åˆ†é’Ÿ
        ]
        
        # åˆ›å»ºæ”¯æŒçƒ­æ›´æ–°çš„æœåŠ¡å™¨
        # âš ï¸ é‡è¦ï¼šä½¿ç”¨ [::] ç›‘å¬æ‰€æœ‰æ¥å£ï¼Œç¡®ä¿ Docker å®¹å™¨é—´å¯ä»¥é€šä¿¡
        # å¦‚æœä½¿ç”¨ localhostï¼Œå…¶ä»–å®¹å™¨æ— æ³•è¿æ¥
        listen_addr = f"[::]:{port}"  # æ”¯æŒ IPv4 å’Œ IPv6
        server, reloader = create_hot_reload_server(
            service_name="auth_service",
            module_path="services.auth_service.grpc_server",
            servicer_class_name="AuthServicer",
            add_servicer_to_server_func=auth_pb2_grpc.add_AuthServiceServicer_to_server,
            port=port,
            server_options=server_options,
            max_workers=20,
            check_interval=30,  # 30ç§’æ£€æŸ¥ä¸€æ¬¡
            listen_addr=listen_addr
        )
        
        # æ³¨å†Œçƒ­æ›´æ–°å™¨
        register_microservice_reloader("auth_service", reloader)
        
        # å¯åŠ¨çƒ­æ›´æ–°ç›‘æ§
        reloader.start()
        
        server.start()
        print(f"ğŸš€ auth-service å¯åŠ¨æˆåŠŸï¼ˆçƒ­æ›´æ–°æ¨¡å¼ï¼‰: {listen_addr}", flush=True)
        
        try:
            server.wait_for_termination()
        except KeyboardInterrupt:
            print("\n>>> æ­£åœ¨åœæ­¢æœåŠ¡...", flush=True)
            reloader.stop()
            server.stop(grace=5)
            print("âœ… æœåŠ¡å·²åœæ­¢", flush=True)
        
    except ImportError:
        # é™çº§åˆ°æ™®é€šæ¨¡å¼
        print("âš ï¸  çƒ­æ›´æ–°æ¨¡å—ä¸å¯ç”¨ï¼Œä½¿ç”¨æ™®é€šæ¨¡å¼", flush=True)
        server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
        auth_pb2_grpc.add_AuthServiceServicer_to_server(AuthServicer(), server)
        
        # âš ï¸ é‡è¦ï¼šä½¿ç”¨ [::] ç›‘å¬æ‰€æœ‰æ¥å£ï¼Œç¡®ä¿ Docker å®¹å™¨é—´å¯ä»¥é€šä¿¡
        listen_addr = f"[::]:{port}"  # æ”¯æŒ IPv4 å’Œ IPv6
        server.add_insecure_port(listen_addr)
        server.start()
        print(f"ğŸš€ auth-service å¯åŠ¨æˆåŠŸï¼ˆæ™®é€šæ¨¡å¼ï¼‰: {listen_addr}", flush=True)
        server.wait_for_termination()
    except Exception as e:
        print(f"âŒ auth-service å¯åŠ¨å¤±è´¥: {e}", flush=True)
        import traceback
        traceback.print_exc()
        raise


if __name__ == "__main__":
    from .config import AUTH_SERVICE_PORT
    serve(port=AUTH_SERVICE_PORT)
